src/bytes.rs

```diff
@@ -861,72 +912,158 @@ impl From<Bytes> for Vec<u8> {
 impl fmt::Debug for Vtable {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Vtable")
+            .field("type_id", &self.type_id)
             .field("clone", &(self.clone as *const ()))
+            .field("will_truncate", &(self.will_truncate as *const ()))
+            .field("into_vec", &(self.into_vec as *const ()))
             .field("drop", &(self.drop as *const ()))
             .finish()
     }
 }
 
 // ===== impl StaticVtable =====
 
-const STATIC_VTABLE: Vtable = Vtable {
-    clone: static_clone,
-    to_vec: static_to_vec,
-    drop: static_drop,
-};
+struct StaticImpl(&'static [u8]);
 
-unsafe fn static_clone(_: &AtomicPtr<()>, ptr: *const u8, len: usize) -> Bytes {
-    let slice = slice::from_raw_parts(ptr, len);
-    Bytes::from_static(slice)
-}
+unsafe impl BytesImpl for StaticImpl {
+    fn into_bytes_parts(this: Self) -> (AtomicPtr<()>, *const u8, usize) {
+        let mut bytes = mem::ManuallyDrop::new(Bytes::from_static(this.0));
```

Comment:
`into_bytes_parts` is used to construct a bytes instance using `with_impl`, correct?
So this constructs a `Bytes` instance in order to decompose the Bytes instance so that we can construct a `Bytes` instance? 
That is a tad confusing. 

Do we even need to have a `StaticImpl`? 
Since `from_static` "manually" constructs a VTable and the Bytes struct, it could just use free method versions of into_bytes_parts, clone, drop, into_vec.

Reply:
I agree things are confusing with `StaticImpl` but it's because `Bytes::from_static()` is `const fn` so we can't use traits within it on stable. And I tried to DRY as much as possible in other places. Maybe it would be better to have right dependencies between functions even if it duplicates some logic?

`StaticImpl` would be needed to allow downcasting `Bytes` to `&'static [u8]` without reinventing the wheel. Actual API for it is not included. I'm not sure which one is better, adding `Bytes::try_into_static()` or `impl TryFrom<Bytes> for &'static [u8]` or exposing `SharedImpl` to public.

Reply:
Ah. I didn't notice the const.   I think further decomposing the logic into utility functions, and then duplicating the minimum amount is the best approach. 

> ... or exposing SharedImpl to public. 

I think the solution here is "layers".  We offer some default implementations, similar to SharedImpl, as a lower-level public API so that people can compose their own higher-level implementations of BytesImpl